### ISSUE TITLE
If node name changes internally, it doesn't change in saved graphs

### ISSUE THREAD

**Issue by Stumblinbear** [newcat/baklavajs#37]
Is this intended? I mean, I guess if you wish to retain renaming, it makes sense, however unless a user has renamed a node, it should be updated on next load to the internal name, don't you think?

**Comment by Stumblinbear** [comment]
Well this extends further than expected. The entire node is saved, then loaded, instead of only saving the values and connections? Any reason you chose this method of persistence?

**Comment by newcat** [comment]
Sorry, I don't completely understand what you mean. Do you mean the node type with internal names? And what do you mean by entire node when it comes to saving and loading?

**Comment by Stumblinbear** [comment]
Well, in SelectOption, for example, the saved node contains the value _as well as_ all of the potential values when saved. Also the display name of the node is saved, and used when loading. I don't persist node names to the database, so when they're re-loaded they have no name. I've fixed this by applying them after loading, however it seems a bit odd.

**Comment by newcat** [comment]
You are right with the select option. Normally, the potential values should not be in the value property. But currently, the only real way to pass values into a node option is by using the `values` field. I am not happy with that as well and will change it. It is, however, a breaking change, so that might need some time.

Regarding the saving of display names: If I understood you correctly, you don't store the full object of the `save()` method, but instead only parts of it, which don't include the `name` property of nodes? The reason for that is that when a user renames a node, the user defined name needs to be stored somewhere - in this case in the name property. When you don't store this property, the name will obviously be lost when loading it again. The way you are doing it by re-applying the name after loading works, but be aware that when your users rename a node, this change will be lost on reload.

**Comment by Stumblinbear** [comment]
Actually, I just replace the name if it's empty on load. However, if I could disallow renaming entirely, I'd prefer that.

I use my own format for saving that uses Objects for interfaces instead of arrays, `[nodeId, interfaceName]` for connections, and doesn't include the name of the node if it wasn't renamed. I do this because it also needs to be loaded by a system written in another language, and that'd make it easier to deal with.

**Comment by florianleux** [comment]
I encountered a similar bug (i think).
I'm adding options to a node within a custom template (extending the Node.vue template) using addOption(), but it seems not to be saved in the editor save.
Do I need to call a method or trigger a specific event to save the new option in the editor ?

**Comment by florianleux** [comment]
Ok I looked a bit more into this bug and it seems that the editor loads only the options declared in the NodeBuilder declaration and ignore any additional option added after (as they are not declared in the node Type class).

It seems that there is no way to change that on my side as in the JS object produced by save, the options of a node are only arrays with label and value.

**Comment by newcat** [comment]
@florianleux Yes, this is intentional since there is a lot of meta-data missing when saving a graph to completely restore all the interfaces / options (e. g. interface types, ...).

You can, however, override the `save` and `load` methods when using the class approach instead of the `NodeBuilder` to implement the functionality yourself. Have a look at the [AdvancedNode](https://github.com/newcat/baklavajs/blob/master/packages/baklavajs-playground/src/AdvancedNode.ts) which does pretty much what you want to do.

**Comment by florianleux** [comment]
Overriding save() and load() worked perfectly !
Thank you very much for your prompt response !
